sw/device/lib/dif/dif_hmac.c

```diff
@@ -0,0 +1,329 @@
+                                          const void *data, size_t len,
+                                          size_t *bytes_sent) {
+  if (hmac == NULL || data == NULL) {
+    return kDifHmacFifoBadArg;
+  }
+
+  const uint8_t *data_sent = (const uint8_t *)data;
+  int bytes_remaining = len;
+
+  while (bytes_remaining > 0 && get_fifo_available_space(hmac) > 0) {
+    bool word_aligned = (uintptr_t)data_sent % sizeof(uint32_t) == 0;
+    int bytes_written = 0;
+
+    if (bytes_remaining < sizeof(uint32_t) || !word_aligned) {
+      // Individual byte writes are needed if the buffer isn't aligned or there
+      // are no more full words to write.
+      mmio_region_write8(hmac->base_addr, HMAC_MSG_FIFO_REG_OFFSET, *data_sent);
+      bytes_written = 1;
+    } else {
+      // |data_sent| is word-aligned and there are still words to write.
+      uint32_t word = read_32(data_sent);
+      mmio_region_write32(hmac->base_addr, HMAC_MSG_FIFO_REG_OFFSET, word);
+      bytes_written = sizeof(uint32_t);
+    }
+
+    bytes_remaining -= bytes_written;
+    data_sent += bytes_written;
+  }
+
+  if (bytes_sent != NULL) {
+    *bytes_sent = len - bytes_remaining;
+  }
+
+  if (bytes_remaining > 0) {
+    return kDifHmacFifoFull;
+  }
+
+  return kDifHmacFifoOk;
+}
+
+dif_hmac_result_t dif_hmac_fifo_count_entries(const dif_hmac_t *hmac,
+                                              uint32_t *num_entries) {
+  if (hmac == NULL || num_entries == NULL) {
+    return kDifHmacBadArg;
+  }
+
+  *num_entries = get_fifo_entry_count(hmac);
+
+  return kDifHmacOk;
+}
+
+dif_hmac_result_t dif_hmac_get_message_length(const dif_hmac_t *hmac,
+                                              uint64_t *msg_len) {
+  if (hmac == NULL || msg_len == NULL) {
+    return kDifHmacBadArg;
+  }
+
+  uint64_t msg_lower =
+      mmio_region_read32(hmac->base_addr, HMAC_MSG_LENGTH_LOWER_REG_OFFSET);
+  uint64_t msg_upper =
+      mmio_region_read32(hmac->base_addr, HMAC_MSG_LENGTH_UPPER_REG_OFFSET);
+
+  *msg_len = (msg_upper << 32) | msg_lower;
+
+  return kDifHmacOk;
+}
+
+dif_hmac_result_t dif_hmac_run(dif_hmac_t *hmac) {
+  if (hmac == NULL) {
+    return kDifHmacBadArg;
+  }
+
+  mmio_region_nonatomic_set_bit32(hmac->base_addr, HMAC_CMD_REG_OFFSET,
+                                  HMAC_CMD_HASH_PROCESS);
+  return kDifHmacOk;
+}
+
+dif_hmac_result_t dif_hmac_done(const dif_hmac_t *hmac, bool *done) {
+  if (hmac == NULL || done == NULL) {
+    return kDifHmacBadArg;
+  }
+
+  // Check if hmac_done is asserted.
+  *done = mmio_region_get_bit32(hmac->base_addr, HMAC_INTR_STATE_REG_OFFSET,
+                                HMAC_INTR_ENABLE_HMAC_DONE);
+  if (*done) {
+    // Clear hmac_done.
+    mmio_region_nonatomic_set_bit32(hmac->base_addr, HMAC_INTR_STATE_REG_OFFSET,
+                                    HMAC_INTR_ENABLE_HMAC_DONE);
+  }
+
+  return kDifHmacOk;
+}
+
+dif_hmac_result_t dif_hmac_digest_read(const dif_hmac_t *hmac,
```

Comment:
Can you merge these functions into a single one that checks for done and populates `digest` if it's complete? The use pattern here seems to be just always calling `_done` and then calling `_digest_read`, so merging them seems sensible.